pkg/storegateway/bucket.go

```diff
@@ -385,10 +372,44 @@ func (s *BucketStore) syncBlocks(ctx context.Context, initialSync bool) error {
 // InitialSync perform blocking sync with extra step at the end to delete locally saved blocks that are no longer
 // present in the bucket. The mismatch of these can only happen between restarts, so we can do that only once per startup.
 func (s *BucketStore) InitialSync(ctx context.Context) error {
-	if err := s.syncBlocks(ctx, true); err != nil {
+	previouslyLoadedBlocks, err := indexheader.RestoreLoadedBlocks(s.dir)
+	if err != nil {
+		level.Warn(s.logger).Log(
+			"msg", "loading the list of index-headers from snapshot file failed; not eagerly loading index-headers for tenant",
+			"dir", s.dir,
+			"err", err,
+		)
+		// Don't fail initialization. If eager loading doesn't happen, then we will load index-headers lazily.
+		// Lazy loading which is slower, but not worth failing startup for.
+	}
```

Comment:
How about we move this chunk inside `s.loadBlocks` (and maybe rename this method to `s.restorePreviousBlocks`)? We don't need to do it if `EagerLoadingStartupEnabled` is disabled. And also this will make the piece about "not failing the initialization" less awkward (IMHO).

Reply:
after we run syncBlocks the snapshotter will be running, so there's a theoretical race between the snapshotter and reading the snapshot from before the shutdown. It's not obvious, so I now added a comment explaining this and moved this whole thing into a function.

But if you have ideas for how to make this better I can do it

Reply:
It makes more sense now, thanks for explaining. I'm ok to stick to the current version. I think, the comments you've added makes it clear 👍